### ISSUE TITLE
[python-package] best_iteration remains 0 when no early stopping is used

### ISSUE THREAD

**Issue by GuiBandeiraCosta** [microsoft/LightGBM#7010]
### Description

When training a model using `lightgbm.train` (see [[engine.py](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py)](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py)), I noticed that the returned booster always has `best_iteration = 0` unless early stopping is enabled.

For example, when I track the best iteration outside of LightGBM’s training loop, I consistently get `best_iteration = 0` regardless of how many boosting rounds were actually run.

---

### Steps to Reproduce

```python
import lightgbm as lgb

# Example dataset
train_data = lgb.Dataset(X_train, label=y_train)
valid_data = lgb.Dataset(X_valid, label=y_valid)

params = {
    "objective": "binary",
    "metric": "binary_logloss",
}

booster = lgb.train(
    params=params,
    train_set=train_data,
    num_boost_round=50,
    valid_sets=[valid_data],
    valid_names=["valid"],
    # No early_stopping_rounds
)

print("Best iteration:", booster.best_iteration)  # Always prints 0
```

---

### Observed Behavior

* `booster.best_iteration` is always `0` unless `early_stopping_round` is set.
* The value only gets updated inside the `except callback.EarlyStopException` block.

Relevant snippet from `engine.py`:

```python
booster.best_iteration = 0
...
except callback.EarlyStopException as earlyStopException:
    booster.best_iteration = earlyStopException.best_iteration + 1
```

This means that if no early stopping is triggered, `best_iteration` never updates.

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

What you've found is intentional behavior... `best_iteration` is an implementation detail of early stopping. What were you hoping it would be set to?

It's safe to unconditionally pass `Booster.best_iteration` through to `predict()` because `0` is interpreted as "all trees".

If you want to get the size of the model, use `.num_trees()` instead of `.best_iteration`.

I'll leave this open because we could do a better job of documenting this. I'll put up some docs improvements.

**Comment by GuiBandeiraCosta** [comment]
Hi james, thanks for answering!
Assuming the variable name of best_iteration I though it would return the iteration that achieved the best score in any given case
Is there anyway to see this when I am not using early stopping? I currently have a aux variable within the engine code to store the best iteration

**Comment by jameslamb** [comment]
>  _the iteration that achieved the best score in any given case_

This sounds simple but it really is not.

Consider how you might choose the "best" iteration in these situations:

* no evaluation metrics tracked during training (this is the default in LightGBM)
* no validation sets supplied
* multiple `metrics` used (how do you pick a single "best" iteration amongst them?)
* multiple validation sets supplied (how do you pick a single "best" iteration? which validation set matters for that?)
* continued training from a previously-trained model (do you re-evaluate validation metrics on all the previous iterations in case the "best" one lies somewhere in the already-pre-trained trees?)

There's more discussion of this in https://github.com/microsoft/LightGBM/pull/6424#discussion_r1581570857

It's for reasons like this that `best_iteration` is only updated when early stopping is used.

> _Is there anyway to see this when I am not using early stopping?_

Yes. You can inspect the evaluation results yourself and choose the "best" iteration in whatever way you'd like. Here's an example:


```python
import lightgbm as lgb
import numpy as np
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=1_000, n_informative=5, n_features=5)

dtrain = lgb.Dataset(X, label=y)

eval_results = dict()
bst = lgb.train(
    train_set=dtrain,
    params={
        "metric": ["rmse", "mape"],
        "objective": "regression",
        "seed": 708,
    },
    valid_sets=[dtrain],
    num_boost_round=100,
    callbacks = [lgb.record_evaluation(eval_results)]
)

# find best iteration on the 'mape' metric
np.argmin(np.array(eval_results["training"]["mape"]))
```

And then when you find the "best" iteration, you can pass that to `predict()` to only use the trees up to that point in generating predictions.

**Comment by GuiBandeiraCosta** [comment]
This makes a lot of sense. I hadn't considered the other possible issues we would face.
Thanks for the detailed answer!

**Comment by jameslamb** [comment]
No problem, thanks for using LightGBM!
